checker/tests/guieffect/Java8Lambdas.java

```diff
@@ -0,0 +1,176 @@
+        public void unsafeDoUI(UIFunctionalInterface<UIElement> func) {
+            //:: error: (call.invalid.ui)
+            func.executeUI(this.arg);
+        }
+
+        public void doEither(@PolyUI PolymorphicFunctionalInterface<UIElement> func) {
+            // In a real program some intelligent dispatch could be done here to avoid running on UI thread unless
+            // needed.
+            arg.runOnUIThread(
+                    new IAsyncUITask() {
+                        final UIElement e2 = arg;
+
+                        @Override
+                        public void doStuff() { // should inherit UI effect
+                            func.executePolymorphic(e2); // should be okay
+                        }
+                    });
+        }
+
+        public void doUISafely(@UI PolymorphicFunctionalInterface<UIElement> func) {
+            // In a real program some intelligent dispatch could be done here to avoid running on UI thread unless
+            // needed.
+            arg.runOnUIThread(
+                    new IAsyncUITask() {
+                        final UIElement e2 = arg;
+
+                        @Override
+                        public void doStuff() { // should inherit UI effect
+                            func.executePolymorphic(e2); // should be okay
+                        }
+                    });
+        }
+    }
+
+    @PolyUIType
+    private static class PolymorphicLambdaRunner {
+        private final UIElement arg;
+
+        public PolymorphicLambdaRunner(UIElement arg) {
+            this.arg = arg;
+        }
+
+        @PolyUIEffect
+        public void doEither(@PolyUI PolymorphicFunctionalInterface<UIElement> func) {
+            func.executePolymorphic(this.arg);
+        }
+    }
+
+    public static void safeContextTestCases(UIElement elem) {
+        LambdaRunner runner = new LambdaRunner(elem);
+        runner.doSafe(e -> e.repaint());
+        //:: error: (call.invalid.ui)
+        runner.doSafe(e -> e.dangerous()); // Not allowed in doSafe
+        //:: error: (call.invalid.ui)
+        runner.doUI(e -> e.repaint()); // Not allowed in safe context
+        //:: error: (call.invalid.ui)
+        runner.doUI(e -> e.dangerous()); // Not allowed in safe context
+        runner.doEither(e -> e.repaint());
+        runner.doEither(e -> e.dangerous());
+        runner.doUISafely(e -> e.dangerous());
+        @AlwaysSafe PolymorphicLambdaRunner safePolymorphicLambdaRunner = new PolymorphicLambdaRunner(elem);
+        safePolymorphicLambdaRunner.doEither(e -> e.repaint());
+        // This next two are ok for this patch since the behavior is the same (no report) for lambdas as for annon
+        // classes. However, shouldn't this be (argument.type.incompatible) just because safePolymorphicLambdaRunner
+        // is not an @UI PolymorphicLambdaRunner ? Or, failing that (call.invalid.ui) since doEither is @PolyUIEffect ?
```

Comment:
@csgordon @wmdietl : Am I misunderstanding something about how polymorphic qualifiers work here? This doesn't seem to be specific to the lambda inference stuff, but just in general I seem to be able to pass an `@UI` argument to a `@PolyUI` formal when the receiver is `@AlwaysSafe`

Reply:
Re: your comment in the code: The effect of invoking ```safePolymorphicLambdaRunner.doEither``` should be (and seems to be) ```@SafeEffect```.

It ought to be argument.type.incompatible as you suggest, but if that lambda is annotated with both ```@UI``` and ```@AlwaysSafe```, that indicates a bug somewhere, as that pairing shouldn't be possible.  With that as the stated type, though, it's possible the framework's ```@PolymorphicQualifier``` support does something funny with the bad input.  @wmdietl do you know what the ```@PolymorphicQualifier``` instantiation code would do with an argument that had multiple qualifiers attached to it?

Reply:
Actually, I wrote this before I figured out the issue with doing bottom-up inference on the tail-end of checking, so this is probably also the same root cause --- the ```super``` call on the method invocation occurs before the lambda is inferred to be ```@UI```, so the framework does subtyping check with an assumed ```@AlwaysSafe``` on the lambda.  As with the others, the ideal fix is to find out the proper way to do inference in a separate pass before checking.

Reply:
This one is still assumed safe. But that seems to be the same behavior as for anonymous classes. Is the lack of error in line 108 expected, for example? (that one is not a lambda and I am adding no annotations there)

Reply:
The short version is it's wrong, but it's my fault, not yours.  I'm working on a clean bug report.